pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/oom/metrics/xml/CycloTest.xml

```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data>
+
+    <code-fragment id="full-example">
+        <![CDATA[
+        public class Complicated {
+          public void example() {
+            int x = 0, y = 1, z = 2, t = 2;
+            boolean a = false, b = true, c = false, d = true;
+
+            if (a && b || b && d) {
+              if (y == z) {
+                x = 2;
+              } else if (y == t && !d) {
+                x = 2;
+              } else {
+                x = 2;
+              }
+            } else if (c && d) {
+              while (z < y) {
+                x = 2;
+              }
+            } else if (a && !b) {
+              for (int n = 0; n < t; n++) {
+                x = 2;
+              }
+            } else {
+              switch (x) {
+              case 1:
+                x = 2;
+                break;
+
+              case 2:
+              case 3:
+                x = 2;
+                break;
+
+              default:
+                x = 2;
+                break;
+              }
+            }
+          }
+        }
+        ]]>
+    </code-fragment>
+
+    <test-code>
```

Comment:
I'll reply to both comments here. The real problem is not that the computation must be forced, it's that the PackageStats contained in `Metrics`, being a singleton, is not reinitialised between every test code parsing (although it is in normal conditions, as the structure is reinitialised at each pmd run). For instance, this causes every dummy class named `Foo` with no package declaration in test code to share the same `ClassStats` object. That really messes up the memoization logic. Given two different test codes with a `Foo` class, if each one defines a method with the same name, then the memoized result of the first code will be spit out for the second (instead of recomputing the result). 

`MetricsForceHook` was a temporary solution, but even then I was obliged to rename some classes of the test codes because the bug still shows up when using ResultOptions. A true solution would be to reset the data structure between each test. I'll look into both features you mentioned to see if they can help, I didn't know they existed. Thanks ! :)